### PR TITLE
[WIP] Adding a 'onKeyDown' handler to the quick pick/inputbox

### PR DESCRIPTION
**Pull request by Chillee**
Adds a function like so for the quick pick box for use for extensions.
`handleKeyDown: (e: KeyboardEvent, value: string) => string`

I decided to listen to handleKeyDown as in many cases (such as for the vim extension), we're trying to listen for keys like `<right>` or `<tab>` that are not captured by onType.

The idea right now is that it returns a string that is then assigned to the current input. I'd like to be able to return quick pick selection options as well, but I thought I'd open a PR first to make sure I'm not doing anything stupid.

Trying to solve https://github.com/Microsoft/vscode/issues/426, https://github.com/Microsoft/vscode/issues/23169, etc.

Specifically for VSCodeVim, I'm trying to fix these issues: https://github.com/VSCodeVim/Vim/issues/745 and https://github.com/VSCodeVim/Vim/issues/779

A couple things that probably need to be done.
1. Test to make sure it works for quick pick as well. I suspect it shouldn't be too difficult... should be done in a similar method to how it's done currently.
2. Modify the vscode.proposed.ts to expose the new interface.
3. Write a sample extension to show how it's used.

@rebornix

### CODE REVIEW

File: src/vs/workbench/browser/parts/quickopen/quickOpenController.ts [hunk 1]
```diff
@@ -172,7 +173,7 @@ export class QuickOpenController extends Component implements IQuickOpenService
 		if (this.pickOpenWidget && this.pickOpenWidget.isVisible()) {
 			this.pickOpenWidget.hide(HideReason.CANCELED);
 		}
-
+		// options.handleKeyDown = (e, value) => {console.log(e, value); return value.trim();};
```
Comment (Chillee) [inline src/vs/workbench/browser/parts/quickopen/quickOpenController.ts]: This is unnecessary, was accidentally left in here as a result for testing.

**Comment by bpasero** [comment]
Closing due to inactivity.

**Comment by bpasero** [comment]
@Chillee wasn't aware you are asking for a review, after all this PR was still marked as `WIP` by you. Are you waiting for a review or are you still working on this?

**Comment by Chillee** [comment]
@bpasero I was waiting on a review to make sure I was touching the right places/taking the right approach. Currently, it works for the use case I opened up this PR for, but there hasn't been much other than that.

To turn this into something I would feel comfortable merging, I still need to update the vscode.d.ts, write some sample tests, and extend this inputbox API to the quickpick list as well. However, most of those things would involve doing things similar to what I did here, so I wanted a quick nod of approval first.

TL;DR I still want to work on it, but I'm waiting for a quick sanity check. Apologies if WIP means something else usually. I've typically also used it to solicit feedback.

**Comment by bpasero** [comment]
Sounds good. @rebornix fyi

**Comment by bpasero** [comment]
@Chillee I do not think that we should push forward a solution where DOM types (`KeyboardEvent`) appear in the API. We typically do not expose any DOM relevant things to the extensions because extensions are not running in an environment where DOM access is possible. 

I think for VIM to enable it for the editor we allow to overwrite what happens when the user types and then the extension is taking full control of the output (on the level of commands). We would probably need something like that for other parts where user input is possible. I think that would include any input box we have throughout the UI. An extension could then hook into the typing in these UI controls and decide what should happen. Does that make sense?

I think a PR for this is a bit unrealistic. In the future please make sure to help out for issues that have the "help wanted" label. Those are usually the ones where we think a PR makes sense 👍

**Comment by bpasero** [comment]
Closing as per my comment.
